components/automate-cli/pkg/verifyserver/services/certificatevalidation/certificatevalidation.go

```diff
@@ -0,0 +1,252 @@
+
+	invalidFormatKeys = strings.TrimSuffix(invalidFormatKeys, ", ")
+	return createCheck(constants.KEY_FORMAT_TITLE, false, "", fmt.Sprintf(constants.KEY_FORMAT_ERROR_MESSAGE, invalidFormatKeys), fmt.Sprintf(constants.KEY_FORMAT_RESOLUTION_MESSAGE, invalidFormatKeys))
+}
+
+func (vc *ValidateCertificateService) validateCertificateAlgorithm(certificates map[string]string) models.Checks {
+	vc.log.Debug("Validating Certificates Hashing Algorithm...")
+	invalidAlgoCerts := ""
+	keys := []string{"Root", "Node", "Admin"}
+
+	for _, key := range keys {
+		cert := certificates[key]
+		certificate, err := decodeAndParseCertificate(cert, key)
+		if err != nil {
+			vc.log.Error(err)
+			invalidAlgoCerts += key
+			invalidAlgoCerts += ", "
+			continue
+		}
+
+		// this is for checking that our certificates are hashed using either of PBE-SHA1-3DES,
+		// RSA (2048), SHA-256 algorithms or not.
+		signatureAlgorithm := certificate.SignatureAlgorithm
+		if (signatureAlgorithm != x509.SHA256WithRSA && signatureAlgorithm != x509.ECDSAWithSHA256) &&
+			(certificate.PublicKeyAlgorithm != x509.RSA || certificate.PublicKey.(*rsa.PublicKey).N.BitLen() != 2048) &&
+			signatureAlgorithm.String() != "PBE-SHA1-3DES" {
+			vc.log.Debugf("%s certificate is not hashed using either of PBE-SHA1-3DES, RSA (2048), SHA-256 algorithms", key)
+			invalidAlgoCerts += key
+			invalidAlgoCerts += ", "
+			continue
+		}
+	}
+
+	if invalidAlgoCerts == "" {
+		vc.log.Debug("All certificates are hashed using either of PBE-SHA1-3DES, RSA (2048), SHA-256 algorithms")
+		return createCheck(constants.CERTIFICATE_ALGORITHM_TITLE, true, constants.CERTIFICATE_ALGORITHM_SUCCESS_MESSAGE, "", "")
+	}
+
+	invalidAlgoCerts = strings.TrimSuffix(invalidAlgoCerts, ", ")
+	return createCheck(constants.CERTIFICATE_ALGORITHM_TITLE, false, "", fmt.Sprintf(constants.CERTIFICATE_ALGORITHM_ERROR_MESSAGE, invalidAlgoCerts), fmt.Sprintf(constants.CERTIFICATE_ALGORITHM_RESOLUTION_MESSAGE, invalidAlgoCerts))
+}
+
+func (vc *ValidateCertificateService) CertificateValidation(req models.CertificateCheckRequest) models.CertificateCheckResponse {
+	var response = models.CertificateCheckResponse{}
+
+	certificates := make(map[string]string)
+	privateKeys := make(map[string]string)
+
+	certificates["Root"] = req.RootCertificate
+	certificates["Node"] = req.NodeCertificate
+	certificates["Admin"] = req.AdminCertificate
+	privateKeys["Node-Key"] = req.PrivateKey
+	privateKeys["Admin-Key"] = req.AdminPrivateKey
+
+	checks := vc.validateCertificateExpiry(certificates)
+	response.Checks = append(response.Checks, checks)
+
+	checks = vc.validateCertificateFormat(certificates)
+	response.Checks = append(response.Checks, checks)
+
+	checks = vc.validateKeyFormat(privateKeys)
+	response.Checks = append(response.Checks, checks)
+
+	checks = vc.validateCertificateAlgorithm(certificates)
+	response.Checks = append(response.Checks, checks)
+
+	flag := true
+	for _, k := range response.Checks {
+		if !k.Passed {
+			flag = false
```

Comment:
Why are we creating a flag variable here?

Reply:
done